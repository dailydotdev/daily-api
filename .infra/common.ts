export interface WorkerArgs {
  enableMessageOrdering?: boolean;
  ackDeadlineSeconds?: number;
  expirationPolicy?: {
    ttl: string;
  };
  deadLetterPolicy?: {
    deadLetterTopic: string;
    maxDeliveryAttempts: number;
  };
}

export interface Worker {
  topic: string;
  subscription: string;
  args?: WorkerArgs;
}

export const digestDeadLetter = 'api.v1.personalized-digest-email-dead-letter';

export const workers: Worker[] = [
  {
    topic: 'views',
    subscription: 'add-views-v2',
  },
  {
    topic: 'user-updated',
    subscription: 'api.user-updated-cio',
  },
  {
    topic: 'user-deleted',
    subscription: 'api.user-deleted-cio',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'comment-upvoted-rep',
  },
  {
    topic: 'comment-upvote-canceled',
    subscription: 'comment-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-scout-matched-slack-v2',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-freeform-images',
  },
  {
    topic: 'api.v1.post-content-edited',
    subscription: 'api.post-edited-freeform-images',
  },
  {
    topic: 'comment-commented',
    subscription: 'comment-commented-slack-message',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-rep',
  },
  {
    topic: 'post-upvote-canceled',
    subscription: 'post-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-downvoted',
    subscription: 'api.post-downvoted-rep',
  },
  {
    topic: 'api.v1.post-downvote-canceled',
    subscription: 'api.post-downvote-canceled-rep',
  },
  {
    topic: 'post-commented',
    subscription: 'api.post-commented-images',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-redis',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-slack-message',
  },
  {
    topic: 'post-commented',
    subscription: 'api.vordr-post-comment-prevented',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-redis',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'post-banned-rep',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.post-deleted-comments-cleanup',
  },
  {
    topic: 'pub-request',
    subscription: 'pub-request-rep',
  },
  {
    topic: 'username-changed',
    subscription: 'username-changed-api',
  },
  {
    topic: 'update-comments',
    subscription: 'update-comments-mention',
  },
  {
    topic: 'api.changes',
    subscription: 'api-cdc',
    args: { enableMessageOrdering: true },
  },
  {
    topic: 'api.v1.cdc-notifications',
    subscription: 'api.cdc-notifications',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-real-time',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-mail',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-push',
  },
  {
    topic: 'api.v1.source-privacy-updated',
    subscription: 'api.source-privacy-updated',
  },
  {
    topic: 'api.v1.content-image-deleted',
    subscription: 'api.delete-cloudinary-image',
  },
  {
    topic: 'api.v1.comment-edited',
    subscription: 'api.comment-edited-images',
  },
  {
    topic: 'api.v1.comment-deleted',
    subscription: 'api.comment-deleted-images',
  },
  // Notifications
  {
    topic: 'community-link-rejected',
    subscription: 'api.community-picks-failed-notification',
  },
  {
    topic: 'community-link-access',
    subscription: 'api.community-picks-granted-notification',
  },
  {
    topic: 'post-commented',
    subscription: 'api.article-new-comment-notification.post-commented',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.article-new-comment-notification.comment-commented',
  },
  {
    topic: 'api.v1.post-bookmark-reminder',
    subscription: 'api.post-bookmark-reminder-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-commented-images',
  },
  {
    topic: 'post-upvoted',
    subscription: 'api.article-upvote-milestone-notification',
  },
  {
    topic: 'send-analytics-report',
    subscription: 'api.article-analytics-notification',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.article-report-approved-notification',
  },
  {
    topic: 'pub-request',
    subscription: 'api.source-request-notification',
  },
  {
    topic: 'api.v1.new-post-mention',
    subscription: 'api.post-mention-notification',
  },
  {
    topic: 'api.v1.new-comment-mention',
    subscription: 'api.comment-mention-notification',
  },
  {
    topic: 'user-reputation-updated',
    subscription: 'api.user-reputation-updated-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-reply-notification',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'api.comment-upvote-milestone-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-notification-v2',
  },
  {
    topic: 'api.v1.member-joined-source',
    subscription: 'api.member-joined-source-notification',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-cio',
  },
  {
    topic: 'api.v1.banner-added',
    subscription: 'api.banner-added',
  },
  {
    topic: 'api.v1.banner-deleted',
    subscription: 'api.banner-deleted',
  },
  {
    topic: 'yggdrasil.v1.content-published',
    subscription: 'api.content-published',
  },
  {
    topic: 'api.v1.source-member-role-changed',
    subscription: 'api.source-member-role-changed-notification',
  },
  {
    topic: 'analytics-api.v1.experiment-allocated',
    subscription: 'api.experiment-allocated',
  },
  {
    topic: 'api.v1.source-created',
    subscription: 'api.source-created-squad-user-action',
  },
  {
    topic: 'api.v1.post-collection-updated',
    subscription: 'api.post-collection-updated-notification',
  },
  {
    topic: 'api.v1.user-readme-updated',
    subscription: 'api.user-readme-images',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-personalized-digest-send-type',
  },
  {
    topic: 'api.v1.comment-downvoted',
    subscription: 'api.comment-downvoted-rep',
  },
  {
    topic: 'api.v1.comment-downvote-canceled',
    subscription: 'api.comment-downvote-canceled-rep',
  },
  {
    topic: 'api.v1.squad-public-request',
    subscription: 'api.v1.squad-public-request-notification',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-updated-cio',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-reset-notification',
  },
  {
    topic: 'api.v1.squad-featured-updated',
    subscription: 'api.squad-featured-updated-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-slack-channel-send',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.vordr-post-prevented',
  },
  {
    topic: 'api.v1.user-company-approved',
    subscription: 'api.user-company-approved-cio',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-user-notification',
  },
  {
    topic: 'api.v1.user-top-reader',
    subscription: 'api.user-top-reader-added',
  }
];

export const personalizedDigestWorkers: Worker[] = [
  {
    topic: 'api.v1.generate-personalized-digest',
    subscription: 'api.personalized-digest-email',
  },
  {
    topic: digestDeadLetter,
    subscription: 'api.personalized-digest-email-dead-letter-log',
  },
];
