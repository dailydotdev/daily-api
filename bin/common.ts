import fastq from 'fastq';
import { ReadStream } from 'fs';

const QUEUE_CONCURRENCY = 1;

export const runInQueueStream = async <T>(
  stream: ReadStream,
  callback: (props: T) => Promise<void>,
  queue = QUEUE_CONCURRENCY,
) => {
  let insertCount = 0;
  const insertQueue = fastq.promise(async (props: T) => {
    await callback(props);

    insertCount += 1;
  }, queue);

  stream.on('data', insertQueue.push);

  await new Promise((resolve, reject) => {
    stream.on('error', reject);
    stream.on('end', resolve);
  });
  await insertQueue.drained();
  console.log('update finished with a total of: ', insertCount);
};
