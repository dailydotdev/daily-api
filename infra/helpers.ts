import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
