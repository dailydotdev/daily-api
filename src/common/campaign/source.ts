import {
  CampaignSource,
  CampaignState,
  CampaignType,
  Post,
  Source,
  type ConnectionManager,
} from '../../entity';

import type { AuthContext } from '../../Context';
import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import {
  ensureSourcePermissions,
  SourcePermissions,
} from '../../schema/sources';
import { ValidationError } from 'apollo-server-errors';
import { randomUUID } from 'crypto';
import { addDays } from 'date-fns';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { usdToCores } from '../number';
import { systemUser, updateFlagsStatement } from '../utils';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationBuilder } from '../../notifications/builder';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import type { NotificationCampaignSourceContext } from '../../notifications';

export const validateSquadBoostPermissions = async (
  ctx: AuthContext,
  sourceId: string,
) => {
  const source = await ensureSourcePermissions(
    ctx,
    sourceId,
    SourcePermissions.BoostSquad,
  );

  if (source.flags.campaignId) {
    throw new ValidationError('Source already has a campaign');
  }

  return source;
};

export const getSourceTags = async (
  con: ConnectionManager,
  sourceId: string,
): Promise<string[]> => {
  const result = await con.getRepository(Post).query<{ tag: string }[]>(
    `
      WITH recent_posts AS (
        SELECT id, "sharedPostId"
        FROM post
        WHERE "sourceId" = $1
        ORDER BY "createdAt" DESC
        LIMIT 30
      )
      SELECT DISTINCT ps.keyword AS tag
      FROM post_keyword ps
      INNER JOIN recent_posts rp
        ON ps."postId" = rp.id
        OR ps."postId" = rp."sharedPostId"
      WHERE ps.status = 'allow'
      LIMIT 30;
    `,
    [sourceId],
  );

  return result.map(({ tag }) => tag.trim()).filter(Boolean);
};

export const startCampaignSource = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value } = args;
  const source = await validateSquadBoostPermissions(ctx, value);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignSource).save(
      manager.getRepository(CampaignSource).create({
        id,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: source.id,
        state: CampaignState.Active,
        endedAt,
        sourceId: source.id,
        type: CampaignType.Squad,
      }),
    );

    const campaignId = campaign.id;
    const last30tags = await getSourceTags(manager, source.id);
    const finalTags = last30tags.length > 3 ? last30tags : []; // when it is 3 or below, we set global targeting

    await skadiApiClientV2.startCampaign(campaign, finalTags);

    await manager
      .getRepository(Source)
      .update(
        { id: source.id },
        { flags: updateFlagsStatement<Source>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Squad Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignSource = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(CampaignSource)
      .update({ id: campaignId }, { state: CampaignState.Cancelled });

    await manager
      .getRepository(Source)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Source>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Squad Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const generateCampaignSquadNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignSourceContext,
) => {
  const { campaign, source, event, user } = ctx;

  return builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`)
    .avatarSource(source);
};

export const generateCampaignSquadEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(CampaignSource)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  const source = await con
    .getRepository(Source)
    .findOneByOrFail({ id: campaign.sourceId });

  return {
    start_date: formatMailDate(campaign.createdAt),
    end_date: formatMailDate(campaign.endedAt),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    source_image: source.image,
    source_handle: source.handle,
    source_name: source.name,
  };
};
