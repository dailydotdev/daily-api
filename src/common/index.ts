export * from './base64';
export * from './cloudinary';
export * from './crypto';
export * from './feedGenerator';
export * from './healthCheck';
export * from './pagination';
export * from './pubsub';
export * from './reputation';
export * from './slack';
export * from './twitter';
export * from './users';
export * from './mailing';
export * from './post';
export * from './links';
export * from './utils';
export * from './typedPubsub';
export * from './personalizedDigest';
export * from './vote';
export * from './feed';
export * from './constants';
