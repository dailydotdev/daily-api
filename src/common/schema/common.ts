import { z } from 'zod';

/**
 * Common pagination arguments schema for cursor-based pagination
 * Use with .extend() to add to your query schemas
 */
export const paginationSchema = z.object({
  /**
   * Paginate after opaque cursor
   */
  after: z.string().optional(),
  /**
   * Number of items to return (first N items)
   */
  first: z.number().max(500).optional().default(100),
});

export type PaginationArgs = z.infer<typeof paginationSchema>;
