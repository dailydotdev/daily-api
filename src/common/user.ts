import { DataSource } from 'typeorm';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';
import { ghostUser } from './index';
import { cancelSubscription } from './paddle';
import type { AuthContext, Context } from '../Context';
import { ForbiddenError } from 'apollo-server-errors';
import { logger } from '../logger';
import { CoresRole } from '../types';
import { remoteConfig } from '../remoteConfig';
import { UserTransaction } from '../entity/user/UserTransaction';
import { SubscriptionProvider, SubscriptionStatus } from './plus';
import { deleteUserResume } from './googleCloud';

export const deleteUser = async (
  con: DataSource,
  userId: string,
  messageId?: string,
) => {
  try {
    const { subscriptionFlags } = await con.getRepository(User).findOneOrFail({
      select: ['subscriptionFlags'],
      where: { id: userId },
    });

    // If the user has a subscription, we need to handle it before deletion
    if (subscriptionFlags?.subscriptionId) {
      if (
        subscriptionFlags?.provider === SubscriptionProvider.AppleStoreKit &&
        subscriptionFlags?.status === SubscriptionStatus.Active
      ) {
        throw new Error(
          'Apple subscriptions are not supported for user deletion',
        );
      }

      if (subscriptionFlags?.provider === SubscriptionProvider.Paddle) {
        const isGifted = !!subscriptionFlags.giftExpirationDate;
        // gifted subscription is a one-time payment hence not considered subscription in Paddle's terms
        if (!isGifted) {
          await cancelSubscription({
            subscriptionId: subscriptionFlags.subscriptionId,
          });
        }
        logger.info(
          {
            provider: SubscriptionProvider.Paddle,
            userId,
            isGifted,
            subscriptionId: subscriptionFlags.subscriptionId,
          },
          'Subscription cancelled user deletion',
        );
      }
    }

    // Delete user's resume if exists
    await deleteUserResume(userId);

    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: ghostUser.id,
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: ghostUser.id });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(UserTransaction).update(
        {
          senderId: userId,
        },
        {
          senderId: ghostUser.id,
        },
      );
      await entityManager.getRepository(UserTransaction).update(
        {
          receiverId: userId,
        },
        {
          receiverId: ghostUser.id,
        },
      );
      await entityManager.getRepository(User).delete(userId);
    });
    logger.info(
      {
        userId,
        messageId,
      },
      'deleted user',
    );
  } catch (err) {
    logger.error(
      {
        userId,
        messageId,
        err,
      },
      'failed to delete user',
    );
    throw err;
  }
};

export /**
 * Function creator that wraps a function with auth protection
 * In this case it checks ctx contains userId but it can be expanded
 * in the future to check more advanced things.
 * This is a DX layer to make sure auth required stuff like cores
 * is protected on the function level.
 *
 * @template Props
 * @template Result
 * @param {(props: Props) => Result} fn
 * @return {Result}
 */
const createAuthProtectedFn = <
  Props extends { ctx: Pick<AuthContext, 'userId'> },
  Result,
>(
  fn: (props: Props) => Result,
) => {
  return (props: Props): Result => {
    if (!props.ctx.userId) {
      throw new ForbiddenError('Auth is required');
    }

    return fn(props);
  };
};

export const getUserCoresRole = ({ region }: { region?: string }) => {
  if (!region) {
    return CoresRole.None;
  }

  const ruleForRegion = remoteConfig.vars.coresRoleRules?.find((rule) =>
    rule.regions.includes(region),
  );

  if (ruleForRegion) {
    return ruleForRegion.role;
  }

  return CoresRole.Creator;
};

export const checkUserCoresAccess = ({
  user,
  requiredRole,
}: {
  user: Pick<User, 'id' | 'coresRole'>;
  requiredRole: CoresRole;
}): boolean => {
  if (!user) {
    return false;
  }

  if (typeof user.coresRole !== 'number') {
    return false;
  }

  if (user.coresRole < requiredRole) {
    return false;
  }

  return true;
};

export const checkCoresAccess = async ({
  ctx,
  userId,
  requiredRole,
}: {
  ctx: Context;
  userId: string;
  requiredRole: CoresRole;
}): Promise<boolean> => {
  const user = await ctx.dataLoader.user.load({
    userId,
    select: ['id', 'coresRole'],
  });

  if (!user) {
    return false;
  }

  return checkUserCoresAccess({ user, requiredRole });
};
