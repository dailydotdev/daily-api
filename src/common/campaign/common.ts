import { ValidationError } from 'apollo-server-errors';
import {
  Campaign,
  CampaignPost,
  CampaignType,
  type ConnectionManager,
} from '../../entity';
import { UserTransaction } from '../../entity/user/UserTransaction';
import { parseBigInt } from '../utils';
import { TransferError } from '../../errors';
import { transferCores, throwUserTransactionError, getBalance } from '../njord';
import type { AuthContext } from '../../Context';
import type { EntityManager } from 'typeorm';
import { CAMPAIGN_VALIDATION_SCHEMA } from '../schema/campaigns';
import { getSourceTags } from './source';
import { generatePostBoostEmail, getPostTags } from './post';
import type { NotificationBuilder } from '../../notifications/builder';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationCampaignContext } from '../../notifications';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { queryReadReplica } from '../queryReadReplica';

export interface StartCampaignArgs {
  value: string;
  duration: number;
  budget: number;
}

export interface StartCampaignMutationArgs {
  ctx: AuthContext;
  args: StartCampaignArgs;
}

export const validateCampaignArgs = (
  args: Pick<StartCampaignArgs, 'budget' | 'duration'>,
) => {
  const result = CAMPAIGN_VALIDATION_SCHEMA.safeParse(args);

  if (result.error) {
    throw new ValidationError(result.error.issues[0].message);
  }
};

interface StartCampaignTransferCoresProps {
  ctx: AuthContext;
  campaignId: string;
  userTransaction: UserTransaction;
  manager: EntityManager;
}

export const startCampaignTransferCores = async ({
  ctx,
  campaignId,
  userTransaction,
  manager,
}: StartCampaignTransferCoresProps) => {
  if (ctx.isTeamMember) {
    return {
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: { amount: (await getBalance({ userId: ctx.userId })).amount },
      },
    };
  }

  try {
    const transfer = await transferCores({
      ctx,
      transaction: userTransaction,
      entityManager: manager,
    });

    return {
      transfer,
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: {
          amount: parseBigInt(transfer.senderBalance?.newBalance),
        },
      },
    };
  } catch (error) {
    if (error instanceof TransferError) {
      await throwUserTransactionError({
        ctx,
        entityManager: manager,
        error,
        transaction: userTransaction,
      });
    }

    throw error;
  }
};

export const stopCampaignTransferCores = async ({
  ctx,
  campaignId,
  userTransaction,
  manager,
}: StartCampaignTransferCoresProps) => {
  if (ctx.isTeamMember) {
    return {
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: { amount: (await getBalance({ userId: ctx.userId })).amount },
      },
    };
  }

  try {
    const transfer = await transferCores({
      ctx,
      transaction: userTransaction,
      entityManager: manager,
    });

    return {
      transfer,
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: {
          amount: parseBigInt(transfer.receiverBalance?.newBalance),
        },
      },
    };
  } catch (error) {
    if (error instanceof TransferError) {
      await throwUserTransactionError({
        ctx,
        entityManager: manager,
        error,
        transaction: userTransaction,
      });
    }

    throw error;
  }
};

export interface StopCampaignProps {
  campaign: Campaign;
  ctx: AuthContext;
}

export const getReferenceTags = (
  con: ConnectionManager,
  type: CampaignType,
  referenceId: string,
) => {
  switch (type) {
    case CampaignType.Post:
      return getPostTags(con, referenceId);
    case CampaignType.Squad:
      return getSourceTags(con, referenceId);
    default:
      throw new ValidationError('Unknown campaign type to estimate reach');
  }
};

export enum CampaignUpdateEvent {
  Started = 'CAMPAIGN_STARTED',
  Completed = 'CAMPAIGN_COMPLETED',
  StatsUpdated = 'STATS_UPDATED',
  BudgetUpdated = 'BUDGET_UPDATED',
}

export interface CampaignCompleted {
  budget: string;
}

export interface CampaignStatsUpdate {
  impressions: number;
  clicks: number;
  unique_users: number;
}

export interface CampaignStateUpdate {
  budget: string; // used budget
}

export interface CampaignUpdateEventArgs {
  campaignId: string;
  event: CampaignUpdateEvent;
  unique_users: number;
  data: CampaignCompleted | CampaignStatsUpdate | CampaignStateUpdate;
  d_update: number;
}

export const generateCampaignCompletedNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignContext,
) => {
  const { campaign, source, event, user } = ctx;

  const nb = builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`);

  switch (campaign.type) {
    case CampaignType.Post:
      return nb.avatarUser(user);
    case CampaignType.Squad:
      if (!source) {
        throw new Error(
          `Can't generate Squad Campaign Notification without the Squad`,
        );
      }
      return nb.avatarSource(source);
    default:
      throw new Error(
        `Unable to generate notification for unknown type: ${campaign.type}`,
      );
  }
};

export const generateCampaignCompletedEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(Campaign)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  switch (campaign.type) {
    case CampaignType.Post:
      return generatePostBoostEmail({
        con,
        postId: campaign.referenceId,
        notification,
        campaign,
      });
    // TODO: MI-1007 - generate email once template id is provided
    default:
      return null;
  }
};

export interface UserCampaignStats {
  impressions: number;
  clicks: number;
  spend: number;
  users: number;
}

export const getUserCampaignStats = async (
  ctx: AuthContext,
): Promise<UserCampaignStats> => {
  const result = await queryReadReplica(ctx.con, ({ queryRunner }) =>
    queryRunner.manager
      .getRepository(CampaignPost)
      .createQueryBuilder('c')
      .select(`SUM(COALESCE((c.flags->>'impressions')::int, 0))`, 'impressions')
      .addSelect(`SUM(COALESCE((c.flags->>'users')::int, 0))`, 'users')
      .addSelect(`SUM(COALESCE((c.flags->>'clicks')::int, 0))`, 'clicks')
      .addSelect(`SUM(COALESCE((c.flags->>'spend')::int, 0))`, 'spend')
      .where(`c."userId" = :user`, { user: ctx.userId })
      .getRawOne(),
  );

  return {
    clicks: result.clicks ?? 0,
    impressions: result.impressions ?? 0,
    users: result.users ?? 0,
    spend: result.spend ?? 0,
  };
};
