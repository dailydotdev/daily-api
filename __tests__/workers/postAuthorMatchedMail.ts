import nock from 'nock';
import { Connection, getConnection } from 'typeorm';
import { expectSuccessfulBackground, saveFixtures } from '../helpers';
import { sendEmail, User as GatewayUser } from '../../src/common';
import worker from '../../src/workers/postAuthorMatchedMail';
import { Post, Source } from '../../src/entity';
import { sourcesFixture } from '../fixture/source';
import { gatewayUsersFixture } from '../fixture/user';

jest.mock('../../src/common/mailing', () => ({
  ...(jest.requireActual('../../src/common/mailing') as Record<
    string,
    unknown
  >),
  sendEmail: jest.fn(),
}));

let con: Connection;

beforeAll(async () => {
  con = await getConnection();
});

beforeEach(async () => {
  jest.resetAllMocks();
  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, Post, [
    {
      id: 'p1',
      shortId: 'sp1',
      title: 'P1',
      url: 'http://p1.com',
      score: 0,
      sourceId: 'a',
      createdAt: new Date(2020, 8, 27),
      image: 'https://daily.dev/image.jpg',
    },
  ]);
});

const mockUsersMe = (user: GatewayUser): nock.Scope =>
  nock(process.env.GATEWAY_URL)
    .get('/v1/users/me')
    .matchHeader('authorization', `Service ${process.env.GATEWAY_SECRET}`)
    .matchHeader('user-id', user.id)
    .matchHeader('logged-in', 'true')
    .reply(200, user);

it('should send post author matched mail', async () => {
  const mockedUsers: GatewayUser[] = [gatewayUsersFixture[0]];
  mockedUsers.forEach(mockUsersMe);
  await expectSuccessfulBackground(worker, {
    postId: 'p1',
    authorId: '1',
  });
  expect(sendEmail).toBeCalledTimes(1);
  expect(jest.mocked(sendEmail).mock.calls[0]).toMatchSnapshot();
});
