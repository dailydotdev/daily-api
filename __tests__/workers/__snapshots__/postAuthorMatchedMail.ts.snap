// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should send post author matched mail 1`] = `
Array [
  Object {
    "asm": Object {
      "groupId": 12850,
    },
    "category": "Notification",
    "dynamicTemplateData": Object {
      "discussion_link": "http://localhost:5002/posts/p1",
      "post_image": "https://daily.dev/image.jpg",
      "post_title": "P1",
      "published_at": "Sep 27, 2020",
      "source_image": "http://image.com/a",
    },
    "from": Object {
      "email": "informer@daily.dev",
      "name": "daily.dev",
    },
    "replyTo": Object {
      "email": "hi@daily.dev",
      "name": "daily.dev",
    },
    "templateId": "d-3d3402ec873640e788f549a0680c40bb",
    "to": "ido@daily.dev",
    "trackingSettings": Object {
      "openTracking": Object {
        "enable": true,
      },
    },
  },
]
`;
